### DISCUSSION TITLE
Component Builder for Maya

### DISCUSSION THREAD

**Discussion by sharktacos** [Autodesk/maya-usd#3340] [Show and tell]
Wanted to share a script I've been hammering away on inspired by the Component Builder in Houdini.

 Export geo to USD with support for:
 - Proxy/render display purposes
 - Payloading of geometry file
 - Extent Hints for Bounding Box display of unloaded payloads
 - Custom metadata with originating Maya file

I followed the [USD WorkingGroup Asset Structure Guidelines ](https://github.com/usd-wg/assets/blob/main/docs/asset-structure-guidelines.md), making the geometry scope name "geo" mandatory, and leaving the naming for the proxy and render scopes open to the user.

```
/apple           (xform) ← Component Kind
    /geo         (scope)
        /render   (scope) ← Purpose: render
        /proxy  (scope) ← Purpose: proxy
```
Here's a screenshot of the interface for the file browser with the Outliner on the left:

<img width="1127" alt="image" src="https://github.com/Autodesk/maya-usd/assets/23034637/3495e77f-0733-4de5-b8c3-15a1f422def8">

and here's the resulting asset file:

```
#usda 1.0
(
    defaultPrim = "apple"
    upAxis = "Y"
)

def Xform "apple" (
    prepend apiSchemas = ["GeomModelAPI"]
    assetInfo = {
        asset identifier = @apple.usda@
        string name = "apple"
    }
    customData = {
        string Exported_from = ""
    }
    kind = "component"
    prepend payload = @./apple_payload.usda@</apple>
)
{
    float3[] extentsHint = [(-1.0000002, -1, -1.0000005), (1, 1, 1.0000001)]

    over "geo"
    {
        over "render"
        {
            uniform token purpose = "render"
        }

        over "proxy"
        {
            uniform token purpose = "proxy"
        }
    }
}

```

If you want to give it a spin you'll need these two files: 
[df_USD_geoExport_UI.mel](https://github.com/sharktacos/VFX-software-prefs/blob/main/Maya/scripts/df_USD_geoExport_UI.mel) 
[df_USD_geoExport.py](https://github.com/sharktacos/VFX-software-prefs/blob/main/Maya/scripts/df_USD_geoExport.py)

Copy these into your Maya scripts folder, and call with the following in the Script Editor (Mel):
```
 source df_USD_geoExport_UI;
 mayaUsd_exportGeo();
```
Usage: Select the root prim to export. Confirm proxy and render groups in file dialog. Enter path and filename for root asset. Export. Will create a root asset file and corresponding geom and payload files.

Would love to get feedback on this from folks, as I'm sure that there's plenty of room for improvement. I'm also hoping it can maybe spark a conversation of how this kind of functionality might make its way eventually into the Maya UI in some form.

**Comment by wallworm** [comment]
Thanks for your scripts. We've been ramping up discussions internally about USD assets and building components. Building up a workflow for building up USD component assets is definitely on our radar.

Moving forward, we'd love to hear general feedback from the community on workflows related to transporting data from DCCs into structured assets. There are so many ways to go about this. One method is as you describe with dialogs to confirm actions. Another workflow we've played with is tagging DCC objects and automatically wrapping those into structures. There are so many ways you could approach this, so getting a consensus on a good general-purpose workflow is what we'd like to find.
